src/backend/commands/pg_profile.c

```diff
@@ -0,0 +1,788 @@
+			dpasswordreusemax = defel;
+		}
+	}
+
+	if (dfailedloginattempts && dfailedloginattempts->arg)
+	{
+		failed_login_attempts = intVal(dfailedloginattempts->arg);
+		if (failed_login_attempts == 0 ||
+			failed_login_attempts < PROFILE_UNLIMITED ||
+			failed_login_attempts > PROFILE_MAX_VALID)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("invalid failed login attempts: %d", failed_login_attempts)));
+	}
+
+	if (dpasswordlocktime && dpasswordlocktime->arg)
+	{
+		password_lock_time = intVal(dpasswordlocktime->arg);
+		if (password_lock_time < PROFILE_UNLIMITED ||
+			password_lock_time > PROFILE_MAX_VALID)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("invalid password lock time: %d", password_lock_time)));
+	}
+
+	if (dpasswordreusemax && dpasswordreusemax->arg)
+	{
+		password_reuse_max = intVal(dpasswordreusemax->arg);
+		if (password_reuse_max < PROFILE_UNLIMITED ||
+			password_reuse_max > PROFILE_MAX_VALID)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("invalid password reuse max: %d", password_reuse_max)));
+	}
+
+	/*
+	 * Check the pg_profile relation to be certain the profile doesn't already
+	 * exist.
+	 */
+	pg_profile_rel = table_open(ProfileRelationId, RowExclusiveLock);
+	pg_profile_dsc = RelationGetDescr(pg_profile_rel);
+
+	if (OidIsValid(get_profile_oid(stmt->profile_name, true)))
+		ereport(ERROR,
+				(errcode(ERRCODE_DUPLICATE_OBJECT),
+				 errmsg("profile \"%s\" already exists",
+	    						stmt->profile_name)));
+
+	/*
+	 * Build a tuple to insert
+	 */
+	MemSet(new_record, 0, sizeof(new_record));
+	MemSet(new_record_nulls, false, sizeof(new_record_nulls));
+
+	new_record[Anum_pg_profile_prfname - 1] =
+		DirectFunctionCall1(namein, CStringGetDatum(stmt->profile_name));
+
+	new_record[Anum_pg_profile_prffailedloginattempts - 1] =
+		Int32GetDatum(failed_login_attempts);
+	new_record[Anum_pg_profile_prfpasswordlocktime - 1] =
+		Int32GetDatum(password_lock_time);
+	new_record[Anum_pg_profile_prfpasswordlifetime - 1] =
+		Int32GetDatum(password_life_time);
+	new_record[Anum_pg_profile_prfpasswordgracetime - 1] =
+		Int32GetDatum(password_grace_time);
+	new_record[Anum_pg_profile_prfpasswordreusetime - 1] =
+		Int32GetDatum(password_reuse_time);
+	new_record[Anum_pg_profile_prfpasswordreusemax - 1] =
+		Int32GetDatum(password_reuse_max);
+	new_record[Anum_pg_profile_prfpasswordallowhashed - 1] =
+		Int32GetDatum(password_allow_hashed);
+	new_record_nulls[Anum_pg_profile_prfpasswordverifyfuncdb -1] =
+		true;
+	new_record_nulls[Anum_pg_profile_prfpasswordverifyfunc - 1] =
+		true;
+
+	/*
+	 * GetNewOidForProfile() / GetNewOrPreassignedOid() will return the
+	 * pre-assigned OID, if any, and error out if there was no pre-assigned
+	 * values in binary upgrade mode.
+	 */
+	{
+		profileid = GetNewOidForProfile(pg_profile_rel, ProfileOidIndexId,
+					    				Anum_pg_profile_oid,
+					    				stmt->profile_name);
+	}
+
```

Comment:
Miss a `if` here? or remote the block and indent

Reply:
In here, follow original PG code.

Reply:
Hm...That's a mistake from GPDB due to merge. In our new codes, don't learn from its faults.